src/Core/SortCursor.h

```diff
@@ -322,93 +330,154 @@ struct SortCursorWithCollation : SortCursorHelper<SortCursorWithCollation>
     }
 };
 
+enum class SortingQueueStrategy
+{
+    Default,
+    Batch
+};
 
-/** Allows to fetch data from multiple sort cursors in sorted order (merging sorted data streams).
-  * TODO: Replace with "Loser Tree", see https://en.wikipedia.org/wiki/K-way_merge_algorithm
-  */
-template <typename Cursor>
-class SortingHeap
+/// Allows to fetch data from multiple sort cursors in sorted order (merging sorted data streams).
+template <typename Cursor, SortingQueueStrategy strategy>
+class SortingQueueImpl
 {
 public:
-    SortingHeap() = default;
+    SortingQueueImpl() = default;
 
     template <typename Cursors>
-    explicit SortingHeap(Cursors & cursors)
+    explicit SortingQueueImpl(Cursors & cursors)
     {
         size_t size = cursors.size();
         queue.reserve(size);
+
         for (size_t i = 0; i < size; ++i)
-            if (!cursors[i].empty())
-                queue.emplace_back(&cursors[i]);
+        {
+            if (cursors[i].empty())
+                continue;
+
+            queue.emplace_back(&cursors[i]);
+        }
+
         std::make_heap(queue.begin(), queue.end());
+
+        if constexpr (strategy == SortingQueueStrategy::Batch)
+        {
+            if (!queue.empty())
+                updateBatchSize();
+        }
     }
 
     bool isValid() const { return !queue.empty(); }
 
-    Cursor & current() { return queue.front(); }
+    Cursor & current() requires (strategy == SortingQueueStrategy::Default)
+    {
+        return queue.front();
+    }
+
+    std::pair<Cursor *, size_t> current() requires (strategy == SortingQueueStrategy::Batch)
```

Comment:
Why pointer, not a reference as for the default mode?

Reply:
Will try to change in follow up refactoring pull request. Not sure if pair can take reference as one of its values.